server/models/procedures.sql

```diff
@@ -1334,4 +1334,152 @@ BEGIN
+  IN is_exit TINYINT(1),
+  IN user_id SMALLINT(5) UNSIGNED
+)
+BEGIN
+  INSERT INTO
+    `stock_movement` (`uuid`, `document_uuid`, `depot_uuid`, `lot_uuid`, `entity_uuid`, `description`, `flux_id`, `date`, `quantity`, `unit_cost`, `is_exit`, `user_id`)
+  VALUES
+    (uuid, document_uuid, depot_uuid, lot_uuid, entity_uuid, description, flux_id, date, quantity, unit_cost, is_exit, user_id);
+END $$
+
+-- This processus inserts records relative to the stock movement in the posting journal
+CREATE PROCEDURE PostPurchase (
+  IN document_uuid BINARY(16),
+  IN date DATETIME,
+  IN enterprise_id SMALLINT(5) UNSIGNED,
+  IN project_id SMALLINT(5) UNSIGNED,
+  IN currency_id TINYINT(3) UNSIGNED,
+  IN user_id SMALLINT(5) UNSIGNED
+)
+BEGIN
+  DECLARE InvalidInventoryAccounts CONDITION FOR SQLSTATE '45006';
+  DECLARE current_fiscal_year_id MEDIUMINT(8) UNSIGNED;
+  DECLARE current_period_id MEDIUMINT(8) UNSIGNED;
+  DECLARE current_exchange_rate DECIMAL(19, 4) UNSIGNED;
+  DECLARE transaction_id VARCHAR(100);
+  DECLARE verify_invalid_accounts SMALLINT(5);
+
+  -- getting the curent fiscal year
+  SET current_fiscal_year_id = (
+    SELECT id FROM fiscal_year AS fy
+    WHERE date BETWEEN fy.start_date AND DATE(ADDDATE(fy.start_date, INTERVAL fy.number_of_months MONTH)) AND fy.enterprise_id = enterprise_id
+  );
+
+  -- getting the period id
+  SET current_period_id = (
+    SELECT id FROM period AS p
+    WHERE DATE(date) BETWEEN DATE(p.start_date) AND DATE(p.end_date) AND p.fiscal_year_id = current_fiscal_year_id
+  );
+
+  -- getting the transaction number
+  SET transaction_id = GenerateTransactionId(project_id);
+
+  CALL PostingJournalErrorHandler(enterprise_id, project_id, current_fiscal_year_id, current_period_id, 1, date);
+
+ -- Check that all every inventory has a stock account and a variation account - if they do not the transaction will be Unbalanced
+  SELECT 
+    COUNT(l.uuid) INTO verify_invalid_accounts
+  FROM 
+    lot AS l
+  JOIN 
+  	stock_movement sm ON sm.lot_uuid = l.uuid
+  JOIN 
+  	inventory i ON i.uuid = l.inventory_uuid
+  JOIN 
+  	inventory_group ig ON ig.uuid = i.group_uuid
+  WHERE 
+  	ig.stock_account IS NULL AND 
+    ig.cogs_account IS NULL AND
+    sm.document_uuid = document_uuid;
+  
+  IF verify_invalid_accounts > 0 THEN
+    SIGNAL InvalidInventoryAccounts
+    SET MESSAGE_TEXT = 'Every inventory should belong to a group with a cogs account and stock account.';
+  END IF;
+
+  -- Debiting stock account, by inserting a record to the posting journal table
+  INSERT INTO posting_journal 
+  (
+    uuid, project_id, fiscal_year_id, period_id, trans_id, trans_date,
+    record_uuid, description, account_id, debit, credit, debit_equiv,
+    credit_equiv, currency_id, origin_id, user_id
+  ) 
+  SELECT
+    HUID(UUID()), project_id, current_fiscal_year_id, current_period_id, transaction_id, 
+    date, p.uuid, sm.description, ig.stock_account, pi.total, 0, pi.total, 0, currency_id,
+    9, user_id
```

Comment:
Where did this 9 come from?

Reply:
In the transaction type table.

Reply:
It would be good to put a variable that describes why this is 9 and not 8 or 10.  For example.
```sql
SOME_RANDOM_ID = 9;

INSERT INTO some_table.some_column SET id = SOME_RANDOM_ID
```

Otherwise, it makes it hard to know why this is 9 and not any other number.

Reply:
I agree, but I am not understanding the code example that you written can you be more clear?
For example : if the code must be a variable, how can you edit the followed statement : 
``` INSERT INTO user (id, name, code) VALUES (1, 'some name', 5)```

Reply:
Like this:
```sql
DECLARE USER_MANAGER_CODE INT;

SET USER_MANAGER_CODE = 5;

INSERT INTO user (id, name, code) VALUES (1, 'some name', USER_MANAGER_CODE);
INSERT INTO user (id, name, code) VALUES (2, 'some name', USER_MANAGER_CODE);
```

In this case, instead of a magic number "5", any other programmers editing the code will know what the number "5" signifies.  And, if a test fails (because someone inserted a new code, making the code for the manager 6), a programmer can easily diagnose and correct the error because they will realize that "5" is no longer what the original programmer intended it to be.

Reply:
Here is an example when the [cost of invoice items](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/server/models/procedures.sql#L1250) is computed and the [here is where it is inserted into a table](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/server/models/procedures.sql#L129).